### ISSUE TITLE
GPU Rules

### ISSUE THREAD

**Issue by willtebbutt** [compintell/Mooncake.jl#278]
Places to look for rules and tests for GPU stuff:
- Zygote.jl tests
- Tracker.jl tests
- Lux.jl integration test
- LuxLib.jl unit tests
- NNlib.jl unit tests
- Flux.jl unit tests

@yebai anything else that you can think of that we should be looking at?

**Comment by kshyatt** [comment]
A couple more rules we need for CUDA arrays:
- [ ] `sort` and friends
- [x] `broadcast`, `mapreduce`
- [ ] LinearAlgebra methods (CUBLAS)
    - [x] ~~LinearAlgebra.mul! / generic_matmatmul!~~ Fixed by #1044
- [x] Whatever is blocking `adjoint` in forward mode (`_new_`?)

Will add more as I find them

**Comment by Technici4n** [comment]
For cuBLAS I wonder if there would be a way to share the rules with the standard BLAS 😅

**Comment by kshyatt** [comment]
There may be but there are some subtleties (which I am happy to discuss)

**Comment by yebai** [comment]
@kshyatt, as discussed, it is helpful to make the above list more comprehensive and specify what is missing to provide reasonable GPU support for linear algebra. 

It's important to be realistic about one's bandwidth and aim for a very high standard within those stated scopes.

**Comment by yebai** [comment]
cc @lkdvos

**Comment by yebai** [comment]
> [!CAUTION]
>  Produced with help from Copilot. 

## What Mooncake currently has (`ext/MooncakeCUDAExt.jl`)

| Rule | Coverage |
|---|---|
| Tangent type system | `CuArray` is its own tangent; `CuPtr`, `CuRefValue`, `CuContext` handled |
| `CuArray(undef, dims...)` | Construction of uninitialised arrays |
| `lgetfield` on `CuArray` | Field access (`.data`, `.dims`) |
| `adjoint`/`transpose` of `CuArray` | Via `_new_` rules |
| `sum(::CuFloatArray)` | Reduction to scalar — **reals only, no function arg** |
| `mul!(C,A,B)` for `CuMaybeComplexArray` | 3-arg GEMM via CUBLAS |
| `@from_rrule` NNlib ops | conv, pool, softmax, bias_act!, etc. |

---

## What's missing (vs. Zygote's CUDA test suite)

### 1. GPU broadcasting — the root gap, blocks most tests

Zygote hooks `broadcasted(::AbstractGPUArrayStyle, f, args...)` → `broadcast_forward`,
a ForwardDiff-Dual forward pass compiled into a CUDA kernel. This handles `x .^ n`,
`log.(x)`, `abs2.(x)`, `cis.(x)`, comparisons, structure preservation (Adjoint, Diagonal),
and all complex math. **Mooncake has no equivalent hook.**

### 2. CPU→GPU transfer (`cu(x)`)

Zygote: `@adjoint Base.convert(T, xs::Array) where {T<:AbstractGPUArray}` with pullback
`convert(Array, Δ)`. Mooncake has no `frule!!/rrule!!` for this.

### 3. `sum(f, ::CuArray)` with arbitrary function

Mooncake only has `sum(::CuFloatArray)`. Zygote rewrites `sum(f, x_gpu)` → `sum(f.(x))`
(goes via broadcasting). Mooncake needs either the same rewrite or an explicit rule.
Also absent for complex CuArrays.

### 4. `map(f, ::CuArray)`

No explicit rule. In CUDA.jl `map` is broadcasting under the hood, so fixing (1) fixes
this transitively.

### 5. `mapreduce`/`reduce` on `CuArray`

No explicit rule. These use CUDA reduction kernels that Mooncake can't trace through.
Need explicit rules (e.g. rewrite `mapreduce(f, +, x)` → `sum(f.(x))`).

### 6. `vcat(::CuArray, scalar)`

No GPU-aware vcat rule. Zygote uses `@allowscalar` for the scalar gradient.

---

## Why Mooncake can't just trace through CUDA.jl

Mooncake differentiates by tracing Julia IR. CUDA kernel lowering (`cufunction`, `@cuda`)
produces code that isn't traceable in the same way. Broadcasting and reductions on
`CuArray` ultimately hit GPU kernels that are opaque to Mooncake's tracer — hence
explicit rules are required for anything that bottoms out in a CUDA kernel.

---

## How `broadcast_forward` works (Zygote's approach)

For `N` input arrays, each element is wrapped as a `ForwardDiff.Dual` with a one-hot
partial vector of length `N`:

```julia
dual(x::Real, i, ::Val{N}) = Dual(x, ntuple(==(i), N))
#                                  ^value  ^partials: [0,…,1,…,0] at position i
```

For complex inputs, real and imaginary parts get separate slots (partials length `2N`).

`dual_function(f).(args...)` broadcasts `f` over Dual-wrapped CuArrays.
**`ForwardDiff.Dual` is just a plain struct of numbers** — CUDA.jl compiles it to a kernel.
The output `CuArray{Dual}` contains both the primal value and all `N` partial derivatives
from a single kernel pass.

The pullback then just multiplies cotangents by the stored partials:

```julia
dargs[i] = unbroadcast(args[i], ȳ .* partials.(out, i))
```

This pullback multiplication is itself a broadcast over the GPU — no CPU involvement.

### Cost vs. benefit

| Property | Detail |
|---|---|
| Single kernel pass | Computes value + all N gradients simultaneously |
| Memory cost | Output holds `N+1` floats per element instead of 1 |
| No tape needed | Partials stored in the output array itself |
| Limitation | Element-wise only — not reductions or irregular access |

---

## Zygote GPU rules location

All GPU-specific AD rules live in **`src/lib/broadcast.jl`** under a `using GPUArraysCore`
block (previously `@require CUDA`):

- `@adjoint broadcasted(::AbstractGPUArrayStyle, f, args...) = broadcast_forward(f, args...)`
- `@adjoint Base.convert(::Type{T}, xs::Array) where {T<:AbstractGPUArray}`
- `@adjoint (::Type{T})(xs::Array) where {T<:AbstractGPUArray}`
- `_pullback(cx, ::typeof(sum), f, xs::AbstractGPUArray)` → rewrites to `sum(f.(xs))`

---

## Potential solutions

### Option A — Mirror Zygote's `broadcast_forward` (recommended, highest impact)

Add to `MooncakeCUDAExt.jl`:

```julia
@is_primitive MinimalCtx Tuple{typeof(broadcasted), AbstractGPUArrayStyle, Any, Vararg}
# frule!!: run dual_function(f).(args...) using ForwardDiff.Dual on CuArrays
# rrule!!: same trick, store Dual output, multiply cotangents by partials in pullback
```

This single rule unblocks broadcasting, `map`, the CUDA complex tests, and (with the sum
rewrite) `sum(f, x)`.

### Option B — Add missing scalar rules individually

- `rrule!!` for `Base.convert(CuArray, ::Array)` — fixes `cu(x)` / GPU movement
- `rrule!!` for `sum(f, ::CuMaybeComplexArray)` — rewrites to `sum(f.(x))`
- `rrule!!` for `mapreduce(f, +, ::CuMaybeComplexArray)` — rewrites to `sum(f.(x))`
- `rrule!!` for `vcat` on mixed `CuArray`/scalar arguments

Lower risk individually but leaves the broadcasting gap open.

### Option C — Upstream to ChainRules

The `broadcast_forward` machinery as a ChainRules `rrule` for `broadcasted` dispatched on
`AbstractGPUArrayStyle`, then imported via `@from_rrule`. Benefits all ChainRules-compatible ADs.

---

## Priority order

1. **GPU broadcasting rule** (Option A) — unlocks ~80% of failing tests transitively
2. **`cu(x)` transfer rule** — small, isolated, easy win
3. **`sum(f, ::CuArray)` + `mapreduce` rewrites** — needed if broadcasting rule isn't a perfect superset
4. **`vcat` + scalar GPU indexing** — niche, lower priority

**Comment by yebai** [comment]
(Mostly) fixed by https://github.com/chalk-lab/Mooncake.jl/pull/1056.
